### PR TITLE
Ng policy manager

### PR DESCRIPTION
**Pull request by kofj**
1. Define ng persist replication policy model.
2. Add ng replication policy CURD methods.
3. Implement ng policy manger.

Signed-off-by: Frank Kung <kfanjian@gmail.com>

### CODE REVIEW

File: src/common/dao/replication_job.go [hunk 1]
```diff
@@ -136,6 +137,114 @@ func AddRepPolicy(policy models.RepPolicy) (int64, error) {
 	return policyID, nil
 }
 
+// AddRepPolicyNg ...
+func AddRepPolicyNg(policy models.RepPolicyNg) (int64, error) {
```
Comment (ywk253100) [inline src/common/dao/replication_job.go]: We'll remove all `ng` at last, so can we rename the function name? We can create a new folder called `ng` and put the functions in the folder.

File: src/common/dao/replication_job.go [hunk 2]
```diff
@@ -136,6 +137,114 @@ func AddRepPolicy(policy models.RepPolicy) (int64, error) {
...
+	o := GetOrmer()
+	sql := `insert into replication_policy_ng (name, description, creator, src_registry_id, src_namespaces,
```
Comment (ywk253100) [inline src/common/dao/replication_job.go]: Use the function provided by the orm rather than the raw SQL

File: src/common/dao/replication_job.go [hunk 3]
```diff
@@ -136,6 +137,114 @@ func AddRepPolicy(policy models.RepPolicy) (int64, error) {
...
+	}
+
+	// WARNING: If cols is null then update all columns.
+	_, err := o.Update(policy, props...)
+
+	return err
+}
+
+// DeleteRepPolicyNg ...
+func DeleteRepPolicyNg(id int64) error {
```
Comment (ywk253100) [inline src/common/dao/replication_job.go]: We need to remove the record from database when one policy is deleted, the changes for master branch is submitted https://github.com/goharbor/harbor/pull/6941

File: src/replication/ng/policy/manager.go [hunk 4]
```diff
@@ -32,3 +120,69 @@ type Manager interface {
...
+	var result []*model.Policy
+	for _, policy := range policies {
+		ply, err := convertFromPersistModel(policy)
+		if err != nil {
+			return 0, nil, err
+		}
+
+		result = append(result, ply)
+	}
+	return int64(len(result)), result, nil
```
Comment (ywk253100) [inline src/replication/ng/policy/manager.go]: The first returning value of the function should be the total count, but the `result` seems just a subset?

**Comment by ywk253100** [comment]
Need UT for the functions

**Comment by kofj** [comment]
> Need UT for the functions

I'll Add UT later.

File: src/common/dao/replication_job.go (hunk 3 above)
Comment (kofj) [reply src/common/dao/replication_job.go]: Hard delete? Why we soft-delete before?

File: src/common/dao/replication_job.go (hunk 1 above)
Comment (kofj) [reply src/common/dao/replication_job.go]: OK

File: src/common/dao/replication_job.go (hunk 2 above)
Comment (kofj) [reply src/common/dao/replication_job.go]: I'll remove raw sql with ORM.But why you use raw sql before?

File: src/replication/ng/policy/manager.go (hunk 4 above)
Comment (kofj) [reply src/replication/ng/policy/manager.go]: ok

File: src/common/dao/replication_job.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kofj) [reply src/common/dao/replication_job.go]: `ng`  suffix will be remove lataer. The question should ask @ywk253100 : Why the file named 'replication_'`job`?

File: src/replication/ng/policy/manager.go [hunk 5]
```diff
@@ -15,9 +15,97 @@
...
+	"encoding/json"
+	"strings"
+	"time"
+
+	"github.com/goharbor/harbor/src/common/dao"
+	persist_models "github.com/goharbor/harbor/src/common/models"
 	"github.com/goharbor/harbor/src/replication/ng/model"
 )
 
+func convertFromPersistModel(policy *persist_models.RepPolicyNg) (*model.Policy, error) {
```
[the remark this answers is not in the document]
Comment (kofj) [reply src/replication/ng/policy/manager.go]: I don't think so, those functions just convert model. Maybe should rename to  `convertFromORM` and `convertToORM`.

File: src/replication/ng/dao/policy.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+func GetPolicies(name, namespace string, page, pageSize int64) ([]*model.RepPolicy, error) {
+	var qs = filteredRepPolicyQuerySeter(name, namespace)
+
+	// Paginate
+	if page > 0 && pageSize > 0 {
+		qs = qs.Limit(pageSize, (page-1)*pageSize)
+	}
+
+	var policies []*model.RepPolicy
+	qs.All(&policies)
```
Comment (ywk253100) [inline src/replication/ng/dao/policy.go]: Handle error?

File: src/replication/ng/dao/policy.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+func GetRepPolicy(id int64) (policy *model.RepPolicy, err error) {
+	err = common_dao.GetOrmer().QueryTable(policy).
+		Filter("id", id).One(policy)
+
+	return
+}
+
+// GetRepPolicyByName return special policy by name.
+func GetRepPolicyByName(name string) (policy *model.RepPolicy, err error) {
+
```
Comment (ywk253100) [inline src/replication/ng/dao/policy.go]: Not implement?

File: src/replication/ng/dao/policy.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+// UpdateRepPolicy TODO: update fields by props
+func UpdateRepPolicy(policy *model.RepPolicy, props ...string) (err error) {
+	var o = common_dao.GetOrmer()
+
+	if policy.ID > 0 {
+		_, err = o.Update(policy, props...)
+	} else {
+		return errors.New("policy id MUST > 0")
+	}
+
```
Comment (ywk253100) [inline src/replication/ng/dao/policy.go]: No logic to update the model.

File: src/replication/ng/policy/manager.go [hunk 9]
```diff
@@ -32,3 +125,86 @@ type Manager interface {
...
+}
+
+// List returns all the policies
+func (m *DefaultManager) List(queries ...*model.PolicyQuery) (total int64, polices []*model.Policy, err error) {
+	// default query parameters
+	var name = ""
+	var namespace = ""
+	var page int64 = 1
+	var pageSize int64 = 15
+
```
Comment (ywk253100) [inline src/replication/ng/policy/manager.go]: The default values for page and size are given on the API level, so the default values for them aren't needed here.

File: src/replication/ng/dao/policy.go [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+func filteredRepPolicyQuerySeter(name, namespace string) orm.QuerySeter {
+	var qs = common_dao.GetOrmer().QueryTable(new(model.RepPolicy))
+
+	var cond = orm.NewCondition()
+	// Pull-based polices in the special namespace.
+	// If dest_registry_id equal 0, dest_namespace is the namespace of Harbor.
+	pullBasedCond := cond.And("dest_registry_id", 0).And("dest_namespace", namespace)
+	// Push-based polices in the special namespace.
+	// If src_registry_id equal 0, src_namespaces only have single namespace of Harbor.
+	pushBasedCond := cond.And("src_registry_id", 0).And("src_namespaces", namespace)
```
Comment (ywk253100) [inline src/replication/ng/dao/policy.go]: Source namespaces should be an array, seems the logic here is not correct

Comment (kofj) [reply src/replication/ng/dao/policy.go]: The namespaces is array in the manager's model, but it is serialize to string stored in database, and split with comma. In push-based mode, only one namespace in the namespaces array.

File: src/replication/ng/dao/policy.go (hunk 7 above)
Comment (kofj) [reply src/replication/ng/dao/policy.go]: Forget to commit...

File: src/replication/ng/dao/policy.go (hunk 6 above)
Comment (kofj) [reply src/replication/ng/dao/policy.go]: fixed

File: src/replication/ng/dao/policy.go (hunk 10 above)
Comment (ywk253100) [reply src/replication/ng/dao/policy.go]: The source namespaces may have more than one element in both pull and push mode.

Comment (kofj) [reply src/replication/ng/dao/policy.go]: Replicate multi-namespaces from harbor to another registry in one policy?I don't think it is a good idea.

Comment (kofj) [reply src/replication/ng/dao/policy.go]: By the sync meeting, we just filter name now.

**Pull request merged**